internal/datastore/common/revisions/optimized.go

```diff
@@ -43,45 +41,71 @@ func (cor *CachedOptimizedRevisions) OptimizedRevision(ctx context.Context) (dat
 	defer span.End()
 
 	localNow := cor.clockFn.Now()
-	lastRevision := cor.lastQuantizedRevision.get()
-	if localNow.Before(lastRevision.validThrough) {
-		log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", lastRevision.validThrough).Msg("returning cached revision")
-		span.AddEvent("returning cached revision")
-		return lastRevision.revision, nil
+
+	cor.Lock()
+
+	// Prune the candidates that have definitely expired
+	var numToDrop uint
+	for _, candidate := range cor.candidates {
+		if candidate.validThrough.Add(cor.maxRevisionStaleness).Before(localNow) {
+			numToDrop++
+		} else {
+			break
+		}
+	}
+	cor.candidates = cor.candidates[numToDrop:]
```

Comment:
did you envision allowing a maxStaleness > 1? 

I imagine typically it will be between 0 and 1, in which case there will only ever be two candidates. We could probably have less locking by just storing two atomic revisions and picking between them.

Reply:
I did write it to be able to support more than two, I start with a `current` and `last`, but in the spirit of writing code that handles 0, 1, or N, I thought I would generalize it.